pyasn1/codec/ber/decoder.py

```diff
@@ -138,27 +207,26 @@ def valueDecoder(self, substrate, asn1Spec,
                      tagSet=None, length=None, state=None,
                      decodeFun=None, substrateFun=None,
                      **options):
-        head, tail = substrate[:length], substrate[length:]
 
         if substrateFun:
             return substrateFun(self._createComponent(
                 asn1Spec, tagSet, noValue, **options), substrate, length)
 
-        if not head:
+        if endOfStream(substrate) or not length:
             raise error.PyAsn1Error('Empty BIT STRING substrate')
 
         if tagSet[0].tagFormat == tag.tagFormatSimple:  # XXX what tag to check?
 
-            trailingBits = oct2int(head[0])
+            trailingBits = ord(substrate.read(1))
             if trailingBits > 7:
                 raise error.PyAsn1Error(
                     'Trailing bits overflow %s' % trailingBits
                 )
 
             value = self.protoComponent.fromOctetString(
-                head[1:], internalFormat=True, padding=trailingBits)
+                substrate.read(length - 1), internalFormat=True, padding=trailingBits)
```

Comment:
Regarding the yielding idea on substrate underrun condition: may be we could do something like this here and everywhere where read is concerned:

```
theBytes = substrate.read(length - 1)

if theBytes is None:  # indicates data starvation in O_NONBLOCK mode
    yield  # yields control back to the caller at this point, the caller can resume it from here

if len(theBytes) < length - 1:
    substrate.seek(-len(theBytes), SEEK_CUR)  # unread incomplete input
    yield

# otherwise go ahead 
```

The overall behavior would depend on the underlying file descriptor mode (blocking vs non-blocking) which is totally in the hands of the caller. What we could probably do is to support both cases internally e.g.:

Blocking mode:

* blocking on read(count)
* yielding when read count is less than desired
* yielding when end-of-stream is not yet reached

Non-blocking mode:

* yielding on input starvation
* yielding when read count is less than desired
* yielding when end-of-stream is not yet reached

At the high level, the decoding would look like this:

```
for asn1Object in decoder(stream):
    if asn1Object is None:
        # incomplete input, wait for it (e.g. via select(stream.fileno)) and retry
        continue
    print('received ', asn1Object)
```

Where `decoder(stream)` would produce an iterable (generator) which would emit ASN.1 structures till input stream is exhausted.

If we go this way, I hope we could still preserve backward compatibility in the `decode` function, while `decodeStream` would exhibit all these shiny new features...

In the next major release we could move `decodeStream` to `decode`, as you've suggested.

WDYT?

Reply:
I must admit that I am not so versed in coroutines, generators, etc. I cannot imagine how this would work or be implemented. If you can do this, I will gladly observe and learn.